sw/device/silicon_creator/lib/drivers/flash_ctrl.c

```diff
@@ -0,0 +1,126 @@
+  if (status == NULL) {
+    return kErrorFlashCtrlInvalidArgument;
+  }
+
+  // Read flash operation status.
+  uint32_t op_status = abs_mmio_read32(kBase + FLASH_CTRL_OP_STATUS_REG_OFFSET);
+  // Read flash controller status.
+  uint32_t fc_status = abs_mmio_read32(kBase + FLASH_CTRL_STATUS_REG_OFFSET);
+
+  // Extract operation status bits.
+  status->done = bitfield_bit32_read(op_status, FLASH_CTRL_OP_STATUS_DONE_BIT);
+  status->err = bitfield_bit32_read(op_status, FLASH_CTRL_OP_STATUS_ERR_BIT);
+
+  // Extract flash controller status bits.
+  status->rd_full =
+      bitfield_bit32_read(fc_status, FLASH_CTRL_STATUS_RD_FULL_BIT);
+  status->rd_empty =
+      bitfield_bit32_read(fc_status, FLASH_CTRL_STATUS_RD_EMPTY_BIT);
+  status->init_wip =
+      bitfield_bit32_read(fc_status, FLASH_CTRL_STATUS_INIT_WIP_BIT);
+
+  return kErrorOk;
+}
+
+rom_error_t flash_ctrl_read_start(uint32_t addr, uint32_t word_count,
+                                  flash_ctrl_region_select_t region) {
+  if (is_busy()) {
+    return kErrorFlashCtrlBusy;
+  }
+
+  // Set the address.
+  abs_mmio_write32(kBase + FLASH_CTRL_ADDR_REG_OFFSET, addr);
+
+  // Set the operation of the transaction: read, program, or erase.
+  uint32_t control_reg_val = bitfield_field32_write(
+      0, FLASH_CTRL_CONTROL_OP_FIELD, FLASH_CTRL_CONTROL_OP_VALUE_READ);
+
+  // Set the partition to info here, and set it back below if we're accessing a
+  // data partition.
+  control_reg_val = bitfield_bit32_write(
+      control_reg_val, FLASH_CTRL_CONTROL_PARTITION_SEL_BIT, 1u);
+
+  // Set the partition and info type.
+  switch (region) {
+    case kFlashCtrlRegionData:
+      // 0 indicates data partition.
+      control_reg_val = bitfield_bit32_write(
+          control_reg_val, FLASH_CTRL_CONTROL_PARTITION_SEL_BIT, 0u);
+      break;
+    case kFlashCtrlRegionInfo0:
+      // The partition is already set to info, so just set the info type.
+      control_reg_val = bitfield_field32_write(
+          control_reg_val, FLASH_CTRL_CONTROL_INFO_SEL_FIELD, 0u);
+      break;
+    case kFlashCtrlRegionInfo1:
+      control_reg_val = bitfield_field32_write(
+          control_reg_val, FLASH_CTRL_CONTROL_INFO_SEL_FIELD, 1u);
+      break;
+    case kFlashCtrlRegionInfo2:
+      control_reg_val = bitfield_field32_write(
+          control_reg_val, FLASH_CTRL_CONTROL_INFO_SEL_FIELD, 2u);
+      break;
+    default:
+      return kErrorFlashCtrlInvalidArgument;
+  }
+
+  // Set the number of words.
+  control_reg_val = bitfield_field32_write(
+      control_reg_val, FLASH_CTRL_CONTROL_NUM_FIELD, word_count);
+
+  // Start the transaction.
+  control_reg_val =
+      bitfield_bit32_write(control_reg_val, FLASH_CTRL_CONTROL_START_BIT, true);
+
+  // Write the configuration.
+  abs_mmio_write32(kBase + FLASH_CTRL_CONTROL_REG_OFFSET, control_reg_val);
+
+  return kErrorOk;
+}
+
+int flash_ctrl_fifo_read(int word_count, uint32_t *data_out) {
```

Comment:
Should `word_count` be of type `size_t`?
Should the return value be `size_t` as well?

Reply:
Yeah probably, fixed